packages/application-shell/src/components/quick-access/butler/butler.tsx

```diff
@@ -0,0 +1,742 @@
+import React, {
+  KeyboardEventHandler,
+  ChangeEventHandler,
+  KeyboardEvent,
+  MouseEventHandler,
+  MouseEvent,
+} from 'react';
+import Fuse from 'fuse.js';
+import { FormattedMessage, useIntl } from 'react-intl';
+import last from 'lodash/last';
+import { css, keyframes, ClassNames } from '@emotion/core';
+import LoadingSpinner from '@commercetools-uikit/loading-spinner';
+import { SearchIcon } from '@commercetools-uikit/icons';
+import { customProperties } from '@commercetools-uikit/design-system';
+import { flattenResults } from '../utils';
+import ButlerCommand from '../butler-command';
+import ButlerContainer from '../butler-container';
+import messages from '../messages';
+import {
+  Command,
+  SearchText,
+  SelectedResult,
+  Stack,
+  HistoryEntry,
+} from '../types';
+
+const isSelectAllCombo = (event: KeyboardEvent<HTMLInputElement>) =>
+  event.key === 'a' &&
+  event.metaKey &&
+  !event.ctrlKey &&
+  !event.altKey &&
+  !event.shiftKey;
+
+const isCloseCombo = (event: KeyboardEvent<HTMLInputElement>) =>
+  event.key === 'Escape' &&
+  !event.metaKey &&
+  !event.ctrlKey &&
+  !event.altKey &&
+  !event.shiftKey;
+
+const getPlatform = () => {
+  if (navigator.appVersion.includes('Win')) return 'windows';
+  if (navigator.appVersion.includes('Mac')) return 'macos';
+  if (navigator.appVersion.includes('X11')) return 'unix';
+  if (navigator.appVersion.includes('Linux')) return 'linux';
+
+  return null;
+};
+
+const hasNewWindowModifier = (
+  event: KeyboardEvent<HTMLElement> | MouseEvent<HTMLElement>
+) => {
+  const platform = getPlatform();
+  switch (platform) {
+    case 'macos':
+      return event.metaKey;
+    default:
+      return event.ctrlKey;
+  }
+};
+
+const shakeAnimation = keyframes`
+  from,
+  to {
+    transform: translate3d(0, 0, 0);
+  }
+
+  14%,
+  42%,
+  70% {
+    transform: translate3d(-3px, 0, 0);
+  }
+
+  28%,
+  56%,
+  84% {
+    transform: translate3d(3px, 0, 0);
+  }
+`;
+
+type State = {
+  hasNetworkError: boolean;
+  isLoading: boolean;
+  searchText: SearchText;
+  selectedResult: SelectedResult;
+  // Used for UX when browsing through history
+  enableHistory: boolean;
+  results: Command[];
+  stack: Stack[];
+};
+type Action =
```

Comment:
This is a perfect example for using a reducer: there are lot of state properties which are updated together most of the time, and most of the logic resides in the reducer.
This means that we can just dispatch an action and the reducer takes care of the rest.